runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -960,6 +960,80 @@ TR::Register *J9::X86::AMD64::TreeEvaluator::conditionalHelperEvaluator(TR::Node
 #endif
 
 
+TR::Register *J9::X86::TreeEvaluator::irdbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR_ASSERT(TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads(), "iReadBarrierEvaluator");
+
+#ifdef OMR_GC_CONCURRENT_SCAVENGER
+
+   auto sourceMR = generateX86MemoryReference(node, cg);
+   auto address = TR::TreeEvaluator::loadMemory(node, sourceMR, TR_RematerializableLoadEffectiveAddress, false, cg);
+   address->setMemRef(sourceMR);
+   sourceMR->decNodeReferenceCounts(cg);
+
+   auto object = cg->allocateRegister();
+   auto load = generateRegMemInstruction(L4RegMem, node, object, generateX86MemoryReference(address, 0, cg), cg);
+   cg->setImplicitExceptionPoint(load);
+
```

Comment:
What does setting as an implicit exception point do exactly ? What is the exception that's possible ?

The other question I had was whether a read barrier opcode should only appear under a node that is at the top of the tree at codegen time (i.e. after trees have been lowered). Looking at some IL from logs with the concurrent scavenger enabled, I did not see it appear only under a treetop and so I was wondering if this was by design.

Reply:
The `irdbari` node may be under a `NULLCHK` and hence this is the implicit NULLCHK point.

There is no restriction or necessity for read barrier to have its own treetop. The discussion following https://github.com/eclipse/openj9/pull/3075/files#r221504019 concluded read barrier is not a GC point.

Reply:
If this is the same node which will be used for field watch, it most certainly will be a GC point in that case.

Reply:
Field watch is irrelevant to this Concurrent Scavenger change. Even it is a GC point for field watch, it is NOT a GC point when field watch is disabled/off. This evaluator does NOT assume whether `irdbari` is a GC point, and should work regardless. It is up to the Optimizer generating the trees, and this evaluator generates instructions for Concurrent Scavenger regardless whether `irdbari` has its own treetop or not. Whether the Optimizer should generate `irdbari` under a dedicate treetop is beyong the scope of this changeset.

Reply:
> 
> 
> The `irdbari` node may be under a `NULLCHK` and hence this is the implicit NULLCHK point.
> 
> There is no restriction or necessity for read barrier to have its own treetop. The discussion following https://github.com/eclipse/openj9/pull/3075/files#r221504019 concluded read barrier is not a GC point.

Is that the right link ?

Reply:
> Is that the right link ?

Yes, the link is correct; but I guess this is a bug in GitHub preventing it jumping to the correct comment. If you click the link, then click "Show conversation" right under "runtime/codert_vm/xnathelp.m4", it should show the correct discussion.